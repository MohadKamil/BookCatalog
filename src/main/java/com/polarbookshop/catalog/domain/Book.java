package com.polarbookshop.catalog.domain;

public record Book(String isbn, String title, String author, double price) {
}
