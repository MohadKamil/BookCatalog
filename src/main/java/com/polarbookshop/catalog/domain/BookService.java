package com.polarbookshop.catalog.domain;

import com.polarbookshop.catalog.domain.exceptions.BookExistsException;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class BookService {

    private final BooksRepository booksRepository;

    public BookService(BooksRepository booksRepository) {
        this.booksRepository = booksRepository;
    }

    public List<Book> getBooks() {
        return booksRepository.getBooks();
    }

    public Optional<Book> getBook(String isbn) {
        return booksRepository.getBook(isbn);
    }

    public void deleteBook(String isbn) {
        booksRepository.deleteBook(isbn);
    }

    public Book addBook(Book book) {
        if(booksRepository.isbnExists(book.isbn())) {
            throw new BookExistsException(book.isbn());
        }

        return saveBook(book);
    }

    public Book editBook(String isbn, Book book) {
        var optionalBook = booksRepository.getBook(isbn);

        if (optionalBook.isEmpty()) {
            return addBook(book);
        }
        else {
            var currentBook = optionalBook.get();

            var updatedBook = new Book(currentBook.isbn(),
                    book.title(), book.author(), book.price());

            return saveBook(updatedBook);
        }
    }

    private Book saveBook(Book book) {
        return booksRepository.saveBook(book);
    }
}

