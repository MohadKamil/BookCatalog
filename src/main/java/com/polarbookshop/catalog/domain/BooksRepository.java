package com.polarbookshop.catalog.domain;

import org.springframework.data.repository.ListCrudRepository;

import java.util.Optional;

public interface BooksRepository extends ListCrudRepository<Book,Long> {

    Optional<Book> findByIsbn(String isbn);
    boolean existsByIsbn(String isbn);
    void deleteByIsbn(String isbn);
}
