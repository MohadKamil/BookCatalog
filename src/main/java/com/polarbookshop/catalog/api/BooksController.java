package com.polarbookshop.catalog.api;

import com.polarbookshop.catalog.domain.Book;
import com.polarbookshop.catalog.domain.BookService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/books")
public class BooksController {

    private final BookService bookService;

    public BooksController(BookService bookService) {
        this.bookService = bookService;
    }

    @GetMapping
    public Iterable<Book> getBooks() {
        return bookService.getBooks();
    }


    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public Book createBook(@Valid @RequestBody Book book) {
        return bookService.addBook(book);
    }

    @GetMapping("{isbn}")
    public ResponseEntity<Book> getBook(@PathVariable String isbn) {
        var optionalBook = bookService.getBook(isbn);
        return ResponseEntity.of(optionalBook);
    }
}
