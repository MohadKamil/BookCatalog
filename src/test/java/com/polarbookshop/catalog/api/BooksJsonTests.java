package com.polarbookshop.catalog.api;

import com.polarbookshop.catalog.domain.Book;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.json.JsonTest;
import org.springframework.boot.test.json.JacksonTester;

@JsonTest
public class BooksJsonTests {

    @Autowired
    private JacksonTester<Book> jacksonTester;

    @Test
    public void testBookSerialization() throws Exception {
        var book = new Book("1","Thinking,Fast and Slow","Daniel Kahneman",40);
        var json = jacksonTester.write(book);

        Approvals.verify(json);
    }
}
