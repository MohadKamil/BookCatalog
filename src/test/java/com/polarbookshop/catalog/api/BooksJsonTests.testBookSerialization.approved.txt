JsonContent {"isbn":"1","title":"Thinking,Fast and Slow","author":"Daniel Kahneman","price":40.0} created from com.polarbookshop.catalog.domain.Book